src/consensus.rs

```diff
@@ -89,7 +98,30 @@ impl<T: Proposition> Consensus<T> {
         signed_vote: SignedVote<T>,
         gen: Generation,
     ) -> Result<VoteResponse<T>> {
-        self.log_signed_vote(&signed_vote);
+        let they_terminated = self
+            .get_super_majority_over_super_majorities(
+                &signed_vote
+                    .unpack_votes()
+                    .into_iter()
+                    .map(|v| v.to_owned())
+                    .collect(),
+            )?
+            .is_some();
+        let we_terminated = self
+            .get_super_majority_over_super_majorities(&self.votes.values().cloned().collect())?
+            .is_some();
+
+        if we_terminated && !they_terminated && self.is_new_vote_from_voter(&signed_vote) {
+            info!("[MBR] Obtained new vote from {} after having already reached termination, sending out broadcast for others to catch up.", signed_vote.voter);
+            let proof_sm_over_sm =
+                self.build_super_majority_vote(self.votes.values().cloned().collect(), gen)?;
+            return Ok(VoteResponse::Broadcast(proof_sm_over_sm));
+        }
+
+        // don't log new votes anymore if we terminated
+        if !we_terminated {
```

Comment:
if we've terminated, do we want to keep running the protocol below :thinking: ?

Can we bail early here? Return our existing decision?

Reply:
Agreed, to avoid code repetition/waste I was wondering can we just put the is_SM_SM block first?